desktop/desktopentry/expand_exec.go

```diff
@@ -0,0 +1,141 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package desktopentry
+
+import (
+	"bytes"
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/snapcore/snapd/strutil/shlex"
+)
+
+func shellQuote(s string) string {
+	return "'" + strings.Replace(s, "'", "'\\''", -1) + "'"
```

Comment:
don't we need to do anything about \ itself?

Reply:
No. backslashes within single quotes are not interpreted as escapes by the shell (and by extension the shlex package we're using). This was based on the helper used in the glib implementation:

https://gitlab.gnome.org/GNOME/glib/-/blob/main/glib/gshell.c#L210-245

Reply:
should we have a test that shows what happens in such case though?

Reply:
I've added a test showing that a variety of special characters (spaces, different quote characters, percent signs and dollar signs) in URIs make it through intact:

```go
		// Special characters within URIs are preserved in the
		// final command line.
		in:   `foo %f`,
		uris: []string{"file:///special%20chars%27%22%25%20%24foo"},
		out:  []string{"foo", `/special chars'"% $foo`},
```